### DISCUSSION TITLE
Warning after upgrade to 1.6.0 - It must be typed with Map<String, ???> but it was typed with Map<Long,String>

### DISCUSSION THREAD

**Discussion by katerynasvitlychna** [mapstruct/mapstruct#3735] [Q&A]
I have the following data:
```
class AggregatedResponseDto{
	private Set<String> orgNames;
	private String uploadedBy;
	...
}

class FileDto{
	FileActorDto uploadedBy;
	...
}
```

and mapper method:

```
@Mapping(target = "orgNames", expression = "java(mapOrgNames(dto, orgIdNameMap))"
@Mapping(target = "uploadedBy", expression = "java(mapContactName(dto.getUploadedBy, contactIdNameMap))"
AggregatedResponseDto convert(FileActorDto dto, Map<Long, String> orgIdNameMap, Map<Long, String> contactIdNameMap);
```

mapOrgNames and mapContactName handles mapping of my Maps to either Set<String> or String.

**version 1.4.2.Final** - no warning was provided for the code above

**version 1.6.0** - warning: The Map parameter "orgIdNameMap" cannot be used for property mapping. It must be typed with Map<String, ???> but it was typed with Map<Long,String>. 

My question is, if mapping of Map<Long, ?> to bean was not supported in earlier versions( [https://github.com/mapstruct/mapstruct/discussions/3148#discussioncomment-5502264](https://github.com/mapstruct/mapstruct/discussions/3148#discussioncomment-5502264)), how come warning was never thrown before? 

What would be the better way to handle this?

I tried `@AfterMapping`, hoping it would do the trick - got the same warning. Working with `@Context` didn't work either, since there can be only one context parameter; in my case, I have two. Adding method that would map Map<Long, String> to Map<String, String> didn't help as well

**Comment by katerynasvitlychna** [comment]
@filiphr  created a separate discussion as was agreed

**Comment by filiphr** [comment]
Thanks @katerynasvitlychna. So this is in since 1.5. Was done as part of #1075. 

It is reported because the parameters are not used explicitly.

The only way to avoid this for now would be to use a wrapper. e.g.

```java
    default AggregatedResponseDto map(FileActorDto dto, Map<Long, String> orgIdNameMap, Map<Long, String> contactIdNameMap) {
        return map(new SourceWrapper(dto, orgIdNameMap, contactIdNameMap));
    }

    @Mapping(target = "orgNames", source = "wrapper")
    @Mapping(target = "uploadedBy", source = "wrapper")
    AggregatedResponseDto map(SourceWrapper wrapper);
```

**Reply by csisy** [reply]
Is there any plan on changing this behaviour? Is there any configuration option currently available?

**Reply by filiphr** [reply]
@csisy the warning might go away with #3861.

>Working with `@Context` didn't work either, since there can be only one context parameter; in my case, I have two

perhaps we need to allow having multiple `@Context` parameters. As in your case it fits nicely. In this case you won't even need to use an expression.

So something like the following might be possible mayb

```java
@Mapping(target = "orgNames", source = "dto")
@Mapping(target = "uploadedBy", qualifiedByName = "uploadedBy")
AggregatedResponseDto convert(FileActorDto dto, @Context("organizations") Map<Long, String> orgIdNameMap,  @Context("contacts") Map<Long, String> contactIdNameMap);

default Set<String> mapOrgNames(FileActorDto dto, @Context("organizations") Map<Long, String> orgIdNameMap) {
    // Your own implementation
}

@Named("uploadedBy")
default String mapUploadedBy(String uploadedBy, @Context("contacts") Map<Long, String> contactIdNameMap) {
    // Your own implementation
}
```
